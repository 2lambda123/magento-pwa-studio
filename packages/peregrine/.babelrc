{
  "plugins": [
    "transform-object-rest-spread",
    "transform-class-properties",
    ["transform-react-jsx", { "pragma": "createElement" }],
    ["transform-runtime", {
      "helpers": true,
      "polyfill": false, // polyfills will be handled by preset-env
      "regenerator": false
    }]
  ],
  "presets": [
    ["env", {
      "targets": {
        "browsers": ["> 5%"]
      },
      "modules": false
    }]
  ],
  "env": {
    "test": {
      "plugins": [
        "transform-es2015-modules-commonjs",
        "transform-object-rest-spread",
        "transform-class-properties",
        ["transform-react-jsx", { "pragma": "createElement" }]
      ]
    }
  }
}
