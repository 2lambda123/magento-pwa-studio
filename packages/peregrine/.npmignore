.eslintrc.js
.storybook
/lib/**/__{docs,helpers,mocks,stories,tests}__/**
/scripts
