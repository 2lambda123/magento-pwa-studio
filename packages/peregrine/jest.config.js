module.exports = {
    displayName: 'Peregrine',
    setupFiles: [
        '<rootDir>/scripts/shim.js',
        '<rootDir>/scripts/fetch-mock.js'
    ],
    browser: true,
    testURL: 'https://localhost/',
    coveragePathIgnorePatterns: ['scripts/*', 'node_modules', 'src/index.js'],
    testPathIgnorePatterns: ['dist', 'node_modules']
};
