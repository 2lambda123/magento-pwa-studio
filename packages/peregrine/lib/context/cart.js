import React, { createContext, useContext, useMemo } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import actions from '../store/actions/cart/actions';
import * as asyncActions from '../store/actions/cart/asyncActions';

const CartContext = createContext();

const CartContextProvider = props => {
    const { actions, asyncActions, cartState, children } = props;

    const cartApi = useMemo(
        () => ({
            actions,
            ...asyncActions
        }),
        [actions, asyncActions]
    );

    const contextValue = useMemo(() => [cartState, cartApi], [
        cartApi,
        cartState
    ]);

    return (
        <CartContext.Provider value={contextValue}>
            {children}
        </CartContext.Provider>
    );
};

const mapStateToProps = ({ cart }) => ({ cartState: cart });

const mapDispatchToProps = dispatch => ({
    actions: bindActionCreators(actions, dispatch),
    asyncActions: bindActionCreators(asyncActions, dispatch)
});

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(CartContextProvider);

export const useCartContext = () => useContext(CartContext);
