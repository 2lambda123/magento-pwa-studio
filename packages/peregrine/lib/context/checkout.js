import React, { createContext, useContext, useMemo } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import actions from '../store/actions/checkout/actions';
import * as asyncActions from '../store/actions/checkout/asyncActions';

const CheckoutContext = createContext();

const CheckoutContextProvider = props => {
    const { actions, asyncActions, checkoutState, children } = props;

    const checkoutApi = useMemo(
        () => ({
            actions,
            ...asyncActions
        }),
        [actions, asyncActions]
    );

    const contextValue = useMemo(() => [checkoutState, checkoutApi], [
        checkoutApi,
        checkoutState
    ]);

    return (
        <CheckoutContext.Provider value={contextValue}>
            {children}
        </CheckoutContext.Provider>
    );
};

const mapStateToProps = ({ checkout }) => ({ checkoutState: checkout });

const mapDispatchToProps = dispatch => ({
    actions: bindActionCreators(actions, dispatch),
    asyncActions: bindActionCreators(asyncActions, dispatch)
});

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(CheckoutContextProvider);

export const useCheckoutContext = () => useContext(CheckoutContext);
