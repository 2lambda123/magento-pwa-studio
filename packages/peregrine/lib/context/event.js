import React, { createContext, useContext, useMemo } from "react"
import { Observable } from "@apollo/client"

const EventContext = createContext()
const noop = () => {}

const EventContextProvider = props => {
    const { children } = props
    const eventState = new Observable(noop)

    const eventApi = useMemo(() => ({
        subscribe: (...args) => eventState.subscribe(...args)
    }), [eventState])

    const contextValue = useMemo(() => [eventApi, eventState], [eventApi, eventState])

    return <EventContext.Provider value={contextValue}>{children}</EventContext.Provider>
}

export default EventContextProvider

export const useEventContext = () => useContext(EventContext)
