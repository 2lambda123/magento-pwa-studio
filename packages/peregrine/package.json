{
  "name": "@magento/peregrine",
  "version": "3.0.0",
  "publishConfig": {
    "access": "public"
  },
  "description": "The core runtime of Magento PWA",
  "main": "lib/index.js",
  "scripts": {
    "clean": " ",
    "storybook": "start-storybook -p 9001 -c .storybook",
    "storybook:build": "build-storybook -c .storybook -o storybook-dist"
  },
  "repository": "github:magento-research/pwa-studio",
  "author": "Magento Commerce",
  "license": "(OSL-3.0 OR AFL-3.0)",
  "bugs": {
    "url": "https://github.com/magento-research/pwa-studio/issues"
  },
  "homepage": "https://github.com/magento-research/pwa-studio/tree/master/packages/peregrine#readme",
  "devDependencies": {
    "@storybook/react": "~4.1.13",
    "intl": "~1.2.5",
    "react": "~16.8.3",
    "react-apollo": "~2.5.3",
    "react-dom": "~16.8.3",
    "react-router-dom": "~5.0.0",
    "react-test-renderer": "~16.8.6"
  },
  "peerDependencies": {
    "@babel/runtime": "^7.4.2",
    "react": "^16.8.0",
    "react-apollo": "^2.5.3",
    "react-dom": "^16.8.0",
    "react-router-dom": "~5.0.0"
  },
  "engines": {
    "node": "10.x",
    "yarn": ">=1.12.0"
  },
  "module": "lib/index.js",
  "jsnext:main": "lib/index.js",
  "es2015": "lib/index.js",
  "sideEffects": false
}
