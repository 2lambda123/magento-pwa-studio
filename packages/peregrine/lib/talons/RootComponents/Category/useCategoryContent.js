import { useCallback, useEffect, useState } from 'react';
import { useLazyQuery } from '@apollo/react-hooks';
import { useAppContext } from '@magento/peregrine/lib/context/app';

const DRAWER_NAME = 'filter';
// TODO: This can be replaced by the value from `storeConfig when the PR,
// https://github.com/magento/graphql-ce/pull/650, is released.
const pageSize = 6;
const placeholderItems = Array.from({ length: pageSize }).fill(null);

/**
 * Returns props necessary to render the categoryContent component.
 *
 * @param {object} props.data - The results of a getCategory GraphQL query.
 *
 * @returns {object} result
 * @returns {number} result.categoryId - This category's ID.
 * @returns {string} result.categoryName - This category's name.
 * @returns {object} result.filters - The filters object.
 * @returns {func}   result.handleLoadFilters - A callback function to signal the user's intent to interact with the filters.
 * @returns {func}   result.handleOpenFilters - A callback function that actually opens the filter drawer.
 * @returns {object} result.items - The items in this category.
 * @returns {bool}   result.loadFilters - Whether or not the user has signalled their intent to interact with the filters.
 * @returns {string} result.pageTitle - The text to put in the browser tab for this page.
 */
export const useCategoryContent = props => {
    const {
        categoryId,
        data,
        queries: { getProductFiltersByCategory }
    } = props;

    const [loadFilters, setLoadFilters] = useState(false);
    const [, { toggleDrawer }] = useAppContext();

    const handleLoadFilters = useCallback(() => {
        setLoadFilters(true);
    }, [setLoadFilters]);
    const handleOpenFilters = useCallback(() => {
        setLoadFilters(true);
        toggleDrawer(DRAWER_NAME);
    }, [setLoadFilters, toggleDrawer]);

    const [getFilters, { data: filterData, error: filterError }] = useLazyQuery(
        getProductFiltersByCategory
    );

    useEffect(() => {
        if (filterError) {
            console.error(filterError);
        }
    }, [filterError]);

    useEffect(() => {
        if (categoryId) {
            getFilters({
                variables: {
                    categoryIdFilter: {
                        eq: categoryId
                    }
                }
            });
        }
    }, [categoryId, getFilters]);

    const filters = filterData ? filterData.products.aggregations : null;
    const items = data ? data.products.items : placeholderItems;
    const totalPagesFromData = data
        ? data.products.page_info.total_pages
        : null;
    const categoryName = data ? data.category.name : null;
    const categoryDescription = data ? data.category.description : null;
    // Note: STORE_NAME is injected by Webpack at build time.
    const pageTitle = categoryName
        ? `${categoryName} - ${STORE_NAME}`
        : STORE_NAME;

    return {
        categoryName,
        categoryDescription,
        filters,
        handleLoadFilters,
        handleOpenFilters,
        items,
        loadFilters,
        pageTitle,
        totalPagesFromData
    };
};
