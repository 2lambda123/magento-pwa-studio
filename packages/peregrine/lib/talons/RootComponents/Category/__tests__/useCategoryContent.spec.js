import React from 'react';
import { createTestInstance } from '@magento/peregrine';

import { useCategoryContent } from '../useCategoryContent';
import { useLazyQuery } from '@apollo/client';
import { act } from 'react-test-renderer';

import { useAppContext } from '../../../../context/app';

global.STORE_NAME = 'Venia';

jest.mock('../../../../context/app', () => {
    const state = {};
    const api = {
        actions: { toggleDrawer: jest.fn() }
    };
    const useAppContext = jest.fn(() => [state, api]);

    return { useAppContext };
});

jest.mock('@apollo/client', () => {
    const apolloClient = jest.requireActual('@apollo/client');
    return {
        ...apolloClient,
        useLazyQuery: jest.fn()
    };
});
const Component = props => {
    const talonprops = useCategoryContent(props);

    return <i {...talonprops} />;
};

const mockProps = {
    categoryId: 3,
    data: {
        category: {
            name: 'Jewelry',
            description: 'Jewelry category'
        },
        products: {
            page_info: {
                total_pages: 1
            },
            items: [
                {
                    id: 1,
                    name: 'Ring'
                },
                {
                    id: 2,
                    name: 'Necklace'
                }
            ]
        }
    }
};

const mockProductFiltersByCategoryData = {
    products: {
        aggregations: [
            {
                label: 'Label'
            }
        ]
    }
};

const mockGetFilters = jest.fn();

useLazyQuery.mockReturnValue([
    mockGetFilters,
    { data: mockProductFiltersByCategoryData }
]);

it('returns the proper shape', () => {
    const rendered = createTestInstance(<Component {...mockProps} />);

    const talonProps = rendered.root.findByType('i').props;

    expect(mockGetFilters).toHaveBeenCalled();
    expect(talonProps).toMatchSnapshot();
});

it('sets the filter loading state', () => {
    const rendered = createTestInstance(<Component {...mockProps} />);

    const talonProps = rendered.root.findByType('i').props;

    const { loadFilters, handleLoadFilters } = talonProps;

    expect(loadFilters).toBeFalsy();

    act(() => {
        handleLoadFilters();
    });

    const updatedProps = rendered.root.findByType('i').props;

    expect(updatedProps.loadFilters).toBeTruthy();
});

it('toggles drawer when opening filters', () => {
    const mockToggleDrawer = jest.fn();
    useAppContext.mockReturnValue([
        {},
        {
            toggleDrawer: mockToggleDrawer
        }
    ]);

    const rendered = createTestInstance(<Component {...mockProps} />);

    const talonProps = rendered.root.findByType('i').props;

    const { loadFilters, handleOpenFilters } = talonProps;

    expect(loadFilters).toBeFalsy();

    act(() => {
        handleOpenFilters();
    });

    const updatedProps = rendered.root.findByType('i').props;

    expect(updatedProps.loadFilters).toBeTruthy();
    expect(mockToggleDrawer).toHaveBeenCalled();
});

it('handles default category id', () => {
    const testProps = Object.assign({}, mockProps, {
        categoryId: 0
    });

    createTestInstance(<Component {...testProps} />);

    expect(mockGetFilters).not.toHaveBeenCalled();
});

it('handles no filter data returned', () => {
    useLazyQuery.mockReturnValue([mockGetFilters, {}]);
    const rendered = createTestInstance(<Component {...mockProps} />);

    const talonProps = rendered.root.findByType('i').props;

    expect(talonProps.filters).toBeNull();
});

it('handles no data prop', () => {
    const testProps = {
        categoryId: 0,
        data: null,
        pageSize: 9
    };

    const rendered = createTestInstance(<Component {...testProps} />);

    const talonProps = rendered.root.findByType('i').props;

    expect(talonProps).toMatchSnapshot();
});
