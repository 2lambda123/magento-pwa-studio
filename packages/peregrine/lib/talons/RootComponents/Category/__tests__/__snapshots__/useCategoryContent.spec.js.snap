// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handles no data prop 1`] = `
Object {
  "categoryDescription": null,
  "categoryName": null,
  "filters": null,
  "handleLoadFilters": [Function],
  "handleOpenFilters": [Function],
  "items": Array [
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
  ],
  "loadFilters": false,
  "pageTitle": "Venia",
  "totalPagesFromData": null,
}
`;

exports[`returns the proper shape 1`] = `
Object {
  "categoryDescription": "Jewelry category",
  "categoryName": "Jewelry",
  "filters": Array [
    Object {
      "label": "Label",
    },
  ],
  "handleLoadFilters": [Function],
  "handleOpenFilters": [Function],
  "items": Array [
    Object {
      "id": 1,
      "name": "Ring",
    },
    Object {
      "id": 2,
      "name": "Necklace",
    },
  ],
  "loadFilters": false,
  "pageTitle": "Jewelry - Venia",
  "totalPagesFromData": 1,
}
`;
